### DISCUSSION TITLE
Guidance on MultiApps and Transfers

### DISCUSSION THREAD

**Discussion by robfairh** [idaholab/moose#16051] [Q&A General] [answered]
Hi folks, I am trying to solve a problem that uses two input files and two different meshes. The first input file solves the temperature in a cylinder. The variable changes in 'z' but not in 'r'. I record the temperature along 'z' in a .csv file using a vector-postprocessor. The second file uses the .csv file data to define the values of a BC on a line. For that, I am using 'FunctionDirichletBC' and defining the function values reading the .csv file.

I am trying to define this with the MultiApps and the Transfer actions to run only one simulation instead of two. I haven't used these actions before. Is there any easy way to implement this?

**Comment by aeslaughter** [comment]
I am not sure we have a transfer that exists for this case. I think what you need is a VectorPostprocessor transfer that transfers the entire vector from one system to the other. A vector version of this: [MultiAppPostprocessorTransfer](https://mooseframework.inl.gov/source/transfers/MultiAppPostprocessorTransfer.html).

The best approach might be to add this capability to the existing MultiAppVectorPostprocessorTransfer, I will take a look and see how that might work. It might be easy to add.

**Reply by aeslaughter** [reply]
I am going to try to add this capability quickly, see #16055

**Reply by robfairh** [reply]
Great! Thank you

**Reply by aeslaughter** [reply]
As you noticed in the issue we are going to build something a bit more generic that should work for this, but it will take some time before it is available. Will the answer below work for you?

**Reply by robfairh** [reply]
I'll try it and see if it works!

**Comment by vincentlaboure** [comment, marked as answer]
I'd think the most direct way to accomplish this is using a `LayeredAverage` which you can transfer with a `MultiAppUserObjectTransfer`. Then you might have to create a `VariableDirichletBC` but that should be pretty straightforward.

**Reply by robfairh** [reply]
Okay, I tried this approach, and it doesn't work.
The 'VariableDirichletBC' is the tricky part. As I would be imposing the values of a variable with the values of an auxiliary variable.
And we cannot couple elemental variables into nodal objects.
Can we?

**Reply by robfairh** [reply]
Maybe I am doing something else wrong.
In my sub file, I define the UserObjects action like this:

```
[UserObjects]
  [./sub_app_temp]
    type = LayeredAverage
    direction = y
    variable = temp
    num_layers = 100
    execute_on = linear
    block = 'cool'
  [../]
[]
```
And in my master file, I define the MultiApps and the Transfers actions like this:
```
[MultiApps]
  [./sub]
    type = FullSolveMultiApp
    input_files = simple.i
    no_backup_and_restore = true
  [../]
[]

[Transfers]
  [layered_transfer_from_sub_app]
    type = MultiAppUserObjectTransfer
    direction = from_multiapp
    user_object = sub_app_temp
    variable = sub_app_var
    multi_app = sub
    # displaced_source_mesh = true
    # skip_bounding_box_check = true
  []
[]
```
which defines an AuxVariable like this:
```
[AuxVariables]
  [./sub_app_var]
    order = CONSTANT
    family = MONOMIAL
    block = 'cool'
    initial_condition = 1
  [../]
[]
```

The transfer of the LayeredAverage object is working, and it modifies the AuxVariable. But defining the BC with the AuxVariable I haven't figured that out. Am I missing something obvious here?

**Reply by vincentlaboure** [reply]
Can you try to model your `VariableDirichletBC` on `PenaltyDirichletBC` to impose your boundary condition weakly instead of strongly? The problem with a `NodalBC` (such as `DirichletBC`) is that it does not know what to do when the variable you impose is discontinuous, as in your case. On the other hand, an `IntegratedBC` should work fine.

**Reply by robfairh** [reply]
I'll try that. Thank you!

**Reply by robfairh** [reply]
It worked! Thank you so much @vincentlaboure and @aeslaughter for your time. @aeslaughter I will still be willing to try that new capability under development.

I had to play a little bit around with the value of ```penalty``` to get it to work. Here is a figure with my result: 
![auxvardirbc](https://user-images.githubusercontent.com/43791150/99103685-91bf5880-25a5-11eb-8f2b-25b8eb25983e.png)

The transfer sets the values of the AuxVariable ```sub_app_var``` and then, ```AuxDirichletBC``` imposes those values to the variable ```temp```.

Here are the [.h](https://github.com/idaholab/moose/files/5538461/AuxDirichletBC.txt) and the [.C](https://github.com/idaholab/moose/files/5538466/AuxDirichletBC.txt) just in case someone in the future wants to do something similar.

**Reply by aeslaughter** [reply]
The new transfer capability was just finished off a few days ago: https://mooseframework.inl.gov/source/transfers/MultiAppReporterTransfer.html
